languages/rust/oso/src/oso.rs

```diff
@@ -62,6 +80,53 @@ impl Oso {
         }
     }
 
+    /// Get the actions actor is allowed to take on resource.
+    /// Returns a [std::collections::HashSet] of actions, typed according the return value.
+    /// # Examples
+    /// ```ignore
+    /// oso.load_str(r#"allow(actor: Actor{name: "sally"}, action, resource: Widget{id: 1}) if
+    ///               action in ["CREATE", "READ"];"#);
+    ///
+    /// // get a HashSet of oso::Actions
+    /// let actions: HashSet<Action> = oso.get_allowed_actions(actor, resource)?;
+    ///
+    /// // or Strings
+    /// let actions: HashSet<String> = oso.get_allowed_actions(actor, resource)?;
+    /// ```
+    pub fn get_allowed_actions<Actor, Resource, T>(
+        &self,
+        actor: Actor,
+        resource: Resource,
+    ) -> crate::Result<HashSet<T>>
+    where
+        Actor: ToPolar,
+        Resource: ToPolar,
+        T: FromPolar + Eq + Hash,
+    {
+        let mut query = self
+            .query_rule(
+                "allow",
+                (actor, PolarValue::Variable("action".to_owned()), resource),
+            )
+            .unwrap();
+
+        let mut set = HashSet::new();
+        loop {
+            match query.next() {
```

Comment:
Could write this as a combinator instead.

Maybe `query.try_fold(HashSet::new(), ...` ?

Would save you needing to deal with the Some/None case.

Reply:
I'll give it a whirl.

Reply:
Hey @joshrotenberg! Are you going to make this change, or do you want me to do a final review as is? I don't think this is a must do. More of a personal preference

Reply:
Sorry, I haven't had a chance to look at it, busy few days. If you're ok with the current implementation lets just go with it (and I can fix any minor issues if you have more) and then possibly revisit.